text/0011-fluent-in-prefs-design-review.md

```diff
@@ -0,0 +1,61 @@
+# Fluent in Prefs Design Review
+
+## Introduction
+
+[Fluent](http://projectfluent.org/) is a "family of localization libraries designed to improve how software is translated." Mozilla's localization engineering team (the "engineering team") is migrating Firefox's localization architecture from DTD/properties files to FTL files processed by Fluent, which will enable richer and more accurate translations, runtime locale loading/switching, runtime LTR/RTL and pseudo-locale switching, and a fallback locale. The first target for this migration is the preferences interface (the about:preferences page and its various subdialogs).
+
+The engineering team met with a set of architects and domain experts (the "review team") to discuss risks and mitigations, which constituted a lightweight "design review" of the project's architecture. This document describes the concerns that the review team raised and the risk mitigations that the teams discussed.
+
+## Why Preferences
+
+The team chose preferences because they're text-heavy, include users' names (which exercises Fluent's support for grammatical gender), and because the precision of text matters in preferences to ensure users clearly understand the implications of changes. So preferences are a good way to prove the change to Fluent before adopting it more widely.
+
+Fluent is a big change to the localization process, and the team plans to validate the change by translating just four strings initially (the "name" strings in brand.dtd/properties?) and then verifying those changes across the localization toolchain.
+
+## Chrome Privileges
+
+The review team asked whether there are parts of preferences that don't have chrome privileges, and if that influences the use of Fluent. It seems like the only such part is the Firefox Accounts image (???), which loads over a network and is localized independently of Firefox, so it won't be an issue.
+
+## Search Incompatibility
+
+The search feature of about:preferences is an issue, as the way that it searches preference labels and descriptions is not compatible with the way that Fluent injects text into the page.  The teams discussed options and concluded that there is a straightforward way to update the search implementation to be compatible with Fluent.
+
+Specifically, FTL messages support arbitrary attributes, so an FTL message for a searchable preference control could include a "searchterms" attribute containing the text to search, i.e.
+
+```
+key1
+  .label = "click"
+  .searchterms = " … "
+```
+
+Attribute values can include references to other strings within the FTL file, so localizers could avoid string duplication if desired.  This model would also enable localizers to customize search terms to include extra keywords that aren't included in other strings related to the control.  Finally, this model supports English fallback for searches via the creation of a secondary *document.l10n* object with the `en` locale.
```

Comment:
`(...) Finally, this model supports adding an English fallback for searches via the creation of a secondary *Localization* object with the same resources registered for the "en" locale.`